libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.hpp

```diff
@@ -0,0 +1,269 @@
+        protoboard<FieldT> &pb,
+        const linear_combination<FieldT> &input,
+        const pb_variable<FieldT> &output,
+        const std::string &annotation_prefix);
+
+    void generate_r1cs_constraints();
+    void generate_r1cs_witness();
+};
+
+/// Compute y = x^5
+/// x: input
+/// y: output
+template<typename ppT>
+class flystel_E_power_five_gadget : public gadget<libff::Fr<ppT>>
+{
+    using FieldT = libff::Fr<ppT>;
+
+private:
+    // internal (i.e. intermediate) variables
+    const pb_variable<FieldT> a0;
+    const pb_variable<FieldT> a1;
+
+public:
+    /// input/output
+    const linear_combination<FieldT> input;
+    const pb_variable<FieldT> output;
+
+    flystel_E_power_five_gadget(
+        protoboard<FieldT> &pb,
+        const linear_combination<FieldT> &input,
+        const pb_variable<FieldT> &output,
+        const std::string &annotation_prefix);
+
+    void generate_r1cs_constraints();
+    void generate_r1cs_witness();
+};
+
+/// Compute y = x^1/5
+/// x: input
+/// y: output
+template<typename ppT, class parameters = anemoi_parameters<libff::Fr<ppT>>>
+class flystel_E_root_five_gadget : public gadget<libff::Fr<ppT>>
+{
+    using FieldT = libff::Fr<ppT>;
+
+private:
+    // internal (i.e. intermediate) variables
+    const pb_variable<FieldT> a0;
+    const pb_variable<FieldT> a1;
+
+public:
+    /// input/output
+    const linear_combination<FieldT> input;
+    const pb_variable<FieldT> output;
+
+    flystel_E_root_five_gadget(
+        protoboard<FieldT> &pb,
+        const linear_combination<FieldT> &input,
+        const pb_variable<FieldT> &output,
+        const std::string &annotation_prefix);
+
+    void generate_r1cs_constraints();
+    void generate_r1cs_witness();
+};
+
+/// Anemoi closed Flystel component for fields of prime characteristic
+///
+/// x0,x1: input (x,y in the paper)
+/// y0,y1: output (u,v in the paper)
+///
+/// The component performs the following computation:
+///
+/// a0 = (beta x1^2 + gamma) = Q_gamma(x1)
+/// a1 = (x0 - a0)^{1/alpha} = E_root_five(x0-a0)
+/// a2 = beta (x1-a1)^2 + delta = Q_delta(x1-a1)
+/// y0 = x0 - a0 + a2
+/// y1 = x1 - a1
+///
+/// \note: in [BBCPSVW22] (x0,x1)->(y0,y1) is denoted with (x,y)->(u,v)
+template<typename ppT, class parameters = anemoi_parameters<libff::Fr<ppT>>>
+class flystel_prime_field_gadget : public gadget<libff::Fr<ppT>>
+{
+    using FieldT = libff::Fr<ppT>;
+
+private:
+    // internal (i.e. intermediate) variables
+    const pb_variable<FieldT> a0;
+    const pb_variable<FieldT> a1;
+    const pb_variable<FieldT> a2;
+
+public:
+    const linear_combination<FieldT> input_x0;
+    const linear_combination<FieldT> input_x1;
+    const pb_variable<FieldT> output_y0;
+    const pb_variable<FieldT> output_y1;
+
+    flystel_Q_gamma_prime_field_gadget<ppT, parameters> Q_gamma;
+    flystel_Q_delta_prime_field_gadget<ppT, parameters> Q_delta;
+    flystel_E_root_five_gadget<ppT, parameters> E_root_five;
```

Comment:
Presumably `five` here actually depends on the value in `parameters`, doesn't it?  If so, we need an issue to address this.

Reply:
`five` is the value of $\alpha$ in the $E$ transformation. Its value is independent of the curve (and hence, of the `anemoi_parameters` class). For binary fields, the value is fixed to $\alpha = 3$. For prime fields $\alpha \in {3,5,7,9,11\}$. For prime fields, the choice is up to the user and represents a trade-off between security and efficiency. E.g. higher values of $\alpha$ result in stronger resistance against certain attacks (algebraic attacks, in particular), but are also computationally more expensive. We consider $\alpha=5$ to be a resonable compromise, so all POC implementations use this value ATM.

Reply:
I see.  So, if I've understood correctly, it makes sense to add a static assert in the `flystel_E_root_five_gadget` that `parameters::alpha == 5`.  Is that valid?

Reply:
Yes. Added.